master/custom/factories.py

```diff
@@ -577,3 +584,252 @@ class WindowsARM64ReleaseBuild(WindowsARM64Build):
+                workdir=oot_host_path,
+            )
+        )
+        if self.can_execute_python:
+            self.addStep(
+                ShellCommand(
+                    name="pythoninfo",
+                    description="pythoninfo",
+                    command=["make", "pythoninfo"],
+                    warnOnFailure=True,
+                    env=self.test_environ,
+                    workdir=oot_host_path,
+                )
+            )
+            self.addStep(
+                Test(
+                    command=test,
+                    timeout=self.test_timeout,
+                    usePTY=test_with_PTY,
+                    env=self.test_environ,
+                    workdir=oot_host_path,
+                )
+            )
+            if branch not in ("3",) and "-R" not in self.testFlags:
+                filename = os.path.join(oot_host_path, "test-results.xml")
+                self.addStep(UploadTestResults(branch, filename=filename))
+        self.addStep(
+            Clean(
+                name="Clean build Python",
+                workdir=oot_build_path,
+            )
+        )
+        self.addStep(
+            Clean(
+                name="Clean host Python",
+                workdir=oot_host_path,
+            )
+        )
+
+
+class Wasm32EmscriptenBuild(UnixCrossBuild):
+    """wasm32-emscripten builder
+
+    * Emscripten SDK >= 3.1.12 must be installed
+    * ccache must be installed
+    * Emscripten PATHs must be pre-pended to PATH
+    * ``which node`` must be equal $EMSDK_NODE
+    """
+    factory_tags = ["wasm", "emscripten"]
+    compile_environ = {
+        "CONFIG_SITE": "../../Tools/wasm/config.site-wasm32-emscripten",
+        "EM_COMPILER_WRAPPER": "ccache",
+    }
+
+    host = "wasm32-unknown-emscripten"
+    host_configure_cmd = ["emconfigure", "../../configure"]
+    host_make_cmd = ["emmake", "make"]
+
+
+class Wasm32EmscriptenNodeBuild(Wasm32EmscriptenBuild):
+    buildersuffix = ".emscripten-node"
+    extra_configure_flags = [
+        "--with-emscripten-target=node",
+        "--disable-wasm-dynamic-linking",
+        "--enable-wasm-pthreads",
+    ]
+
+
+class Wasm32EmscriptenBrowserBuild(Wasm32EmscriptenBuild):
+    buildersuffix = ".emscripten-browser"
+    extra_configure_flags = [
+        "--with-emscripten-target=browser",
+        "--enable-wasm-dynamic-linking",
+        "--disable-wasm-pthreads",
+    ]
+    # browser builds do not accept argv from CLI
+    can_execute_python = False
+
+
+class Wasm32WASIBuild(UnixCrossBuild):
+    """wasm32-wasi builder
+
+    * WASI SDK >= 16 must be installed to default path /opt/wasi-sdk
+    * WASIX must be installed to /opt/wasix
+    * ccache must be installed
+    * wasmtime must be installed and on PATH
+    """
+    buildersuffix = ".wasi"
+    factory_tags = ["wasm", "wasi"]
+    extra_configure_flags = [
+        # debug builds exhaust the limited call stack on WASI
+        "--without-pydebug",
+        "--disable-ipv6",
```

Comment:
Why do you disable IPv6? Would you mind to add a comment?

Reply:
It's documented in CPython upstream ``Tools/wasm`` directory. tl;dr WASI does not have networking in traditional sense.

Reply:
Just add a comment in that case.